[GEOS-11725] Environment parameters resolving is not working on Smart data loader

fernandor777 opened the pull request:
[![GEOS-11725](https://badgen.net/badge/JIRA/GEOS-11725/0052CC)](https://osgeo-org.atlassian.net/browse/GEOS-11725) [<img width="16" alt="Powered by Pull Request Badge" src="https://user-images.githubusercontent.com/1393946/111216524-d2bb8e00-85d4-11eb-821b-ed4c00989c02.png">](https://pullrequestbadge.com/?utm_medium=github&utm_source=geoserver&utm_campaign=badge_info)<!-- PR-BADGE: PLEASE DO NOT REMOVE THIS COMMENT -->

JIRA issue:
https://osgeo-org.atlassian.net/browse/GEOS-11725

When a GeoServer catalog have a Postgis datastore working with environment parametrization enabled and parametrizaed parameters including user and password, selecting this store from smart data loader throws an error trying to use the parametrizaed strings without performing the environment resolving.

This PR targets to fix the root problem, allowing smart data loader to reuse parametrized stores.
  
<!-- Please help our volunteers reviewing this PR by completing the following items. 
Ask in a comment if you have troubles with any of them. -->

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [x] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

<!--Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.-->

The PR will be merged when all the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)), there is a code committer review, and the checklist has been fulfilled.

mprins commented on src/community/smart-data-loader/src/main/java/org/geoserver/smartdataloader/data/SmartDataLoaderDataAccessFactory.java:
```diff
@@ -10,13 +10,7 @@
-import java.util.List;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
```
I'm pretty sure wildcard imports are not allowed, but somehow the QA / PMD workflow has not picked up on this... 

see also: https://github.com/geoserver/geoserver/blob/6c2ffdf81676c624c3216b680df0963a1968bcd8/build/qa/pmd-ruleset.xml#L105-L117

fernandor777 replied:
Hi @mprins thanks for your feedback, fixing it.

mprins reviewed (changes requested):
It seems the PMD profile is not triggered during the QA, while I think it should be?... (thus not failing build on wild-card imports)
Apart from that LGTM.

/cc @aaime

aaime commented:
Wondering if it's related to the recent PMD update?

aaime commented on src/community/smart-data-loader/src/main/java/org/geoserver/smartdataloader/data/SmartDataLoaderDataAccessFactory.java:
```diff
@@ -284,9 +286,25 @@ private File saveMappingDocument(String pathname, String filename, Document mapp
+        return catalog.getDataStore(jdbcDataStoreId);
+    }
+
+    static Map<String, Serializable> resolveParams(Map<String, Serializable> params) {
```
Good intention but done in the wrong place IMHO, stores should not be aware of environment parameterization (or we'd have to make the all aware).

The GeoServer resource pool is [already resolving environment parameters on its own here](https://github.com/geoserver/geoserver/blob/99ae245402fca51e306eaf9817faaa891bf89e73/src/main/src/main/java/org/geoserver/catalog/ResourcePool.java#L806).
Please try to see why it's not working as intended, and fix it instead.

I'd keep the tests class instead.

aaime commented:
Thinking out loud, since this is a community module it can also live with some tech debt, but I'd spend a bit of time at least figuring out why the core env substitution is not working.

fernandor777 commented:
Thanks @aaime indeed I investigated and found the way to make resourcepool perform the env parameter resolving, I refactored the implementation and added a test integrating the env parameter on the database host.

aaime commented on src/community/smart-data-loader/src/main/java/org/geoserver/smartdataloader/data/SmartDataLoaderDataAccessFactory.java:
```diff
@@ -284,9 +285,15 @@ private File saveMappingDocument(String pathname, String filename, Document mapp
+        return catalog.getResourcePool().clone(catalog.getDataStore(jdbcDataStoreId), true);
+    }
+
+    private static Serializable resolveValue(Serializable value, GeoServerEnvironment env) {
```
Does not seem to be used anywhere?

fernandor777 replied:
Right, removed it.  Thanks.

fernandor777 replied:
Holy leftover, batman!

Thanks, removed.

aaime commented on src/community/smart-data-loader/src/test/java/org/geoserver/smartdataloader/data/store/EnvParametrizationStoreTest.java:
```diff
@@ -0,0 +1,142 @@
+
+    @BeforeClass
+    public static void testClassSetup() {
+        System.setProperty(ALLOW_ENV_PARAMETRIZATION, "true");
```
I would also have a test where env parametrization is not enabled, to check there is no substitution

aaime reviewed (commented):
Overall looking good, see comments below

aaime reviewed (approved):
Looks good , thanks!
